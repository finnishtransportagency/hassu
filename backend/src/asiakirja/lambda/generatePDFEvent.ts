import {
  AloituskuulutusPdfOptions,
  CreateHyvaksymisPaatosKuulutusPdfOptions,
  CreateNahtavillaoloKuulutusPdfOptions,
  YleisotilaisuusKutsuPdfOptions,
} from "../asiakirjaTypes";

export type GeneratePDFEvent = {
  createAloituskuulutusPdf?: AloituskuulutusPdfOptions;
  createHyvaksymisPaatosKuulutusPdf?: CreateHyvaksymisPaatosKuulutusPdfOptions;
  createYleisotilaisuusKutsuPdf?: YleisotilaisuusKutsuPdfOptions;
  createNahtavillaoloKuulutusPdf?: CreateNahtavillaoloKuulutusPdfOptions;
};
