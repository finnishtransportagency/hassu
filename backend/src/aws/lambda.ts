import { Lambda } from "@aws-sdk/client-lambda";
import { log } from "../logger";

const lambda: Lambda = new Lambda({});

export async function invokeLambda(functionName: string, synchronousCall: boolean, payload?: string): Promise<string | undefined> {
  try {
    if (synchronousCall) {
      const output = await lambda.invoke({
        FunctionName: functionName,
        Payload: new TextEncoder().encode(payload || "{}"),
      });
      const outputPayload = output.Payload;
      return new TextDecoder("utf-8").decode(outputPayload);
    } else {
      await lambda.invokeAsync({
        FunctionName: functionName,
        InvokeArgs: payload || "{}",
      });
    }
  } catch (e) {
    log.error("invokeLambda", e);
    throw e;
  }
}
