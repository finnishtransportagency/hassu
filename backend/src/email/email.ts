import * as nodemailer from "nodemailer";
import { config } from "../config";
import { log } from "../logger";
import { MailOptions } from "nodemailer/lib/smtp-transport";
import cloneDeep from "lodash/cloneDeep";

const transporter = nodemailer.createTransport({
  port: 465,
  host: "email-smtp.eu-west-1.amazonaws.com",
  secure: true,
  auth: {
    user: config.smtpKeyId,
    pass: config.smtpSecret,
  },
  debug: true,
});

export type EmailOptions = Pick<MailOptions, "to" | "subject" | "text" | "attachments">;

async function sendEmail(options: EmailOptions): Promise<void> {
  if (config.emailsOn !== "true") {
    const { attachments: attachments, ...restOptions } = options;
    log.info("Sähköpostin lähetys kytketty pois päältä", {
      emailOptions: restOptions,
      attachments: attachments?.map((att) => {
        const a = cloneDeep(att);
        a.content = "***test***";
        return a;
      }),
    });
    return;
  }
  try {
    const messageInfo = await transporter.sendMail({
      from: "noreply.hassu@vaylapilvi.fi",
      ...options,
      to: config.emailsTo || options.to,
    });
    log.info("Email lähetetty", messageInfo);
  } catch (e) {
    log.error("Email lähetys epäonnistui", e);
  }
}

export const emailClient = { sendEmail };
