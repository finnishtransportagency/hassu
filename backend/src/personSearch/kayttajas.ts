import { Kayttaja } from "../../../common/graphql/apiModel";

export class Kayttajas {
  kayttajaMap: Record<string, Kayttaja>;

  constructor(kayttajaMap: Record<string, Kayttaja>) {
    this.kayttajaMap = kayttajaMap;
  }

  getKayttajaByUid(uid: string | undefined): Kayttaja | undefined {
    if (uid) {
      return this.kayttajaMap[uid];
    }
  }

  findByEmail(email: string): Kayttaja | undefined {
    for (const kayttaja of Object.values(this.kayttajaMap)) {
      if (kayttaja.email === email) {
        return kayttaja;
      }
    }
  }

  static fromKayttajaList(kayttajas: Kayttaja[]) {
    return new Kayttajas(
      kayttajas.reduce((map, kayttaja) => {
        if (kayttaja.uid) {
          map[kayttaja.uid] = kayttaja;
        }
        return map;
      }, {} as Record<string, Kayttaja>)
    );
  }

  asList() {
    const list = [];
    for (const uid in this.kayttajaMap) {
      if (this.kayttajaMap.hasOwnProperty(uid)) {
        list.push(this.kayttajaMap[uid]);
      }
    }
    return list;
  }

  asMap() {
    return this.kayttajaMap;
  }

  findByText(hakusana: string): Kayttaja[] {
    if (hakusana.length >= 3) {
      return Object.values(this.kayttajaMap).reduce((list, kayttaja) => {
        if (kayttaja.etuNimi.includes(hakusana) || kayttaja.sukuNimi.includes(hakusana)) {
          list.push(kayttaja);
        }
        return list;
      }, [] as Kayttaja[]);
    }
    return [];
  }
}
