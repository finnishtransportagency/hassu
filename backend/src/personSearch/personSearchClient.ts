import { Kayttaja, ProjektiRooli } from "../../../common/graphql/apiModel";
import { DBVaylaUser } from "../database/model/projekti";
import { isAorL } from "../user";
import { mergeKayttaja } from "./personAdapter";
import { personSearchUpdaterClient } from "./personSearchUpdaterClient";
import { s3Cache } from "../cache/s3Cache";
import log from "loglevel";
import { Kayttajas } from "./kayttajas";
import { wrapXrayAsync } from "../aws/xray";

export const S3CACHE_TTL_MILLIS = 15 * 60 * 1000; // 15 min
export enum SearchMode {
  EMAIL,
  UID,
}

export const PERSON_SEARCH_CACHE_KEY = "users.json";

export class PersonSearchClient {
  public async getKayttajas(): Promise<Kayttajas> {
    try {
      return await wrapXrayAsync("getKayttajas", async () => {
        const kayttajaMap: Record<string, Kayttaja> = await s3Cache.get(
          PERSON_SEARCH_CACHE_KEY,
          S3CACHE_TTL_MILLIS,
          async () => {
            personSearchUpdaterClient.triggerUpdate();
          },
          async () => {
            return await personSearchUpdaterClient.readUsersFromSearchUpdaterLambda();
          }
        );
        return new Kayttajas(kayttajaMap);
      });
    } catch (e) {
      log.error("getKayttajas", e);
      throw e;
    }
  }

  public async fillInUserInfoFromUserManagement({
    user: user,
    searchMode,
  }: {
    user: Partial<DBVaylaUser>;
    searchMode: SearchMode;
  }): Promise<DBVaylaUser | undefined> {
    const kayttajas = await this.getKayttajas();
    let account: Kayttaja | undefined;

    if (searchMode === SearchMode.UID) {
      account = kayttajas.getKayttajaByUid(user.kayttajatunnus);
    } else {
      account = kayttajas.findByEmail(user.email);
    }
    if (account) {
      // Projektipaallikko must be either L or A account
      if (user.rooli === ProjektiRooli.PROJEKTIPAALLIKKO && !isAorL(account)) {
        return;
      }
      mergeKayttaja(user, account);
      return user as DBVaylaUser;
    }
    return;
  }
}

export const personSearch = new PersonSearchClient();
