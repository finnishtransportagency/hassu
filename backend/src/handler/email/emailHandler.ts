import { emailClient } from "../../email/email";
import { createAloituskuulutusHyvaksyttavanaEmail } from "../../email/emailTemplates";
import { log } from "../../logger";
import { personSearch } from "../../personSearch/personSearchClient";
import { Kayttaja } from "../../../../common/graphql/apiModel";
import Mail from "nodemailer/lib/mailer";
import { DBProjekti, SahkopostiVastaanottaja } from "../../database/model";
import { config } from "../../config";
import { Readable } from "stream";
import { GetObjectCommand, GetObjectOutput } from "@aws-sdk/client-s3";
import SMTPTransport from "nodemailer/lib/smtp-transport";
import { getS3Client } from "../../aws/client";
import { aloituskuulutusHyvaksyntaEmailSender } from "./AloituskuulutusHyvaksyntaEmailSender";
import { hyvaksymisPaatosHyvaksyntaEmailSender } from "./sendHyvaksymiskuulutusEmails";
import { nahtavillaoloHyvaksyntaEmailSender } from "./sendNahtavillaoloKuulutusEmails";

export async function getFileAttachment(oid: string, key: string): Promise<Mail.Attachment | undefined> {
  log.info("haetaan s3:sta sähköpostiin liitetiedosto", { key });

  const getObjectParams = {
    Bucket: config.yllapitoBucketName,
    Key: `yllapito/tiedostot/projekti/${oid}` + key,
  };
  try {
    const output: GetObjectOutput = await getS3Client().send(new GetObjectCommand(getObjectParams));

    if (output.Body instanceof Readable) {
      let contentType = output.ContentType;
      if (contentType == "null") {
        contentType = undefined;
      }
      return {
        filename: getFilename(key),
        contentDisposition: "attachment",
        contentType: contentType || "application/octet-stream",
        content: output.Body,
      };
    } else {
      log.error("Liitetiedoston sisallossa ongelmia");
    }
  } catch (error) {
    log.error("Virhe liitetiedostojen haussa", { error, getObjectParams });
  }

  return Promise.resolve(undefined);
}

function getFilename(path: string): string {
  return path.substring(path.lastIndexOf("/") + 1);
}

export async function getKayttaja(uid: string): Promise<Kayttaja | undefined> {
  const kayttajas = await personSearch.getKayttajas();
  return kayttajas.getKayttajaByUid(uid);
}

export async function sendWaitingApprovalMail(projekti: DBProjekti): Promise<void> {
  const emailOptions = createAloituskuulutusHyvaksyttavanaEmail(projekti);
  if (emailOptions.to) {
    await emailClient.sendEmail(emailOptions);
  } else {
    log.error("Aloituskuulutukselle ei loytynyt projektipaallikon sahkopostiosoitetta");
  }
}

export function examineEmailSentResults(
  vastaanottaja: SahkopostiVastaanottaja,
  sentMessageInfo: SMTPTransport.SentMessageInfo | undefined,
  aikaleima: string
): void {
  // Sähköpostien lähetyksessä tapahtui virhe
  if (!sentMessageInfo) {
    vastaanottaja.lahetysvirhe = true;
  }
  const email = vastaanottaja.sahkoposti;
  if (sentMessageInfo?.accepted.find((accepted) => accepted == email) || sentMessageInfo?.pending?.find((pending) => pending == email)) {
    vastaanottaja.lahetetty = aikaleima;
    vastaanottaja.messageId = sentMessageInfo?.messageId;
    log.info("Email lähetetty", { sentEmail: email, actualEmailAddress: config.emailsTo });
  }
  if (sentMessageInfo?.rejected.find((rejected) => rejected == email)) {
    log.info("Email lähetysvirhe", { rejectedEmail: email });
    vastaanottaja.lahetysvirhe = true;
  }
}

export async function sendAloitusKuulutusApprovalMailsAndAttachments(oid: string): Promise<void> {
  await aloituskuulutusHyvaksyntaEmailSender.sendEmails(oid);
}

export async function sendHyvaksymiskuulutusApprovalMailsAndAttachments(oid: string): Promise<void> {
  await hyvaksymisPaatosHyvaksyntaEmailSender.sendEmails(oid);
}

export async function sendNahtavillaKuulutusApprovalMailsAndAttachments(oid: string): Promise<void> {
  await nahtavillaoloHyvaksyntaEmailSender.sendEmails(oid);
}
