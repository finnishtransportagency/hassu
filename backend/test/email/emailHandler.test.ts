import { describe, it } from "mocha";
import * as sinon from "sinon";
import { TilasiirtymaToiminto } from "../../../common/graphql/apiModel";
import { projektiDatabase } from "../../src/database/projektiDatabase";
import { emailClient } from "../../src/email/email";
import { emailHandler } from "../../src/handler/emailHandler";
import { Kayttajas } from "../../src/personSearch/kayttajas";
import { personSearch } from "../../src/personSearch/personSearchClient";
import { ProjektiFixture } from "../fixture/projektiFixture";
import { PersonSearchFixture } from "../personSearch/lambda/personSearchFixture";
import AWSMock from "aws-sdk-mock";
import AWS from "aws-sdk";

const { expect } = require("chai");

describe("emailHandler", () => {
  afterEach(() => {
    sinon.reset();
    sinon.restore();
    AWSMock.restore();
  });

  beforeEach(() => {
    AWSMock.setSDKInstance(AWS);
  });

  describe("sendEmailsByToiminto", () => {
    let s3Stub: sinon.SinonStub;
    let sendEmailStub: sinon.SinonStub;
    let getKayttajasStub: sinon.SinonStub;
    let loadProjektiByOidStub: sinon.SinonStub;
    let updateAloitusKuulutusJulkaisuStub: sinon.SinonStub;
    let fixture: ProjektiFixture;
    let personSearchFixture: PersonSearchFixture;

    beforeEach(() => {
      s3Stub = sinon.stub();
      sendEmailStub = sinon.stub(emailClient, "sendEmail");
      getKayttajasStub = sinon.stub(personSearch, "getKayttajas");
      loadProjektiByOidStub = sinon.stub(projektiDatabase, "loadProjektiByOid");
      updateAloitusKuulutusJulkaisuStub = sinon.stub(projektiDatabase, "updateAloitusKuulutusJulkaisu");
      fixture = new ProjektiFixture();
      personSearchFixture = new PersonSearchFixture();

      getKayttajasStub.resolves(
        Kayttajas.fromKayttajaList([
          personSearchFixture.pekkaProjari,
          personSearchFixture.mattiMeikalainen,
          personSearchFixture.manuMuokkaaja,
        ])
      );
      sendEmailStub.resolves();
      loadProjektiByOidStub.resolves(fixture.dbProjekti2);
    });

    describe("sendWaitingApprovalMail", () => {
      it("should send email to projektipaallikko succesfully", async () => {
        await emailHandler.sendEmailsByToiminto(TilasiirtymaToiminto.LAHETA_HYVAKSYTTAVAKSI, fixture.PROJEKTI2_OID);

        sinon.assert.calledOnce(sendEmailStub);
        sinon.assert.calledWith(sendEmailStub, {
          subject: "Valtion liikenneväylien suunnittelu: Aloituskuulutus odottaa hyväksyntää ",
          text:
            "Valtion liikenneväylien suunnittelu -järjestelmän projektistasi\n" +
            "Testiprojekti 2 email lahetys\n" +
            "on luotu aloituskuulutus, joka odottaa hyväksyntääsi.\n" +
            "Voit tarkastella projektia osoitteessa https://localhost:3000/yllapito/projekti/2\n" +
            "Saat tämän viestin, koska sinut on merkitty projektin projektipäälliköksi. Tämä on automaattinen sähköposti, johon ei voi vastata.",
          to: "pekka.projari@vayla.fi",
        });
      });
    });

    describe("sendApprovalMailsAndAttachments", () => {
      it("should send emails and attachments succesfully", async () => {
        updateAloitusKuulutusJulkaisuStub.resolves();
        AWSMock.mock("S3", "getObject", s3Stub);
        s3Stub.resolves({});

        await emailHandler.sendEmailsByToiminto(TilasiirtymaToiminto.HYVAKSY, fixture.PROJEKTI2_OID);

        const calls = s3Stub.getCalls();
        expect(calls).to.have.length(2);
        expect(
          calls.map((call) => {
            return { ...call.args[0] };
          })
        ).toMatchSnapshot();
        sinon.assert.callCount(sendEmailStub, 3);
      });
    });
  });
});
