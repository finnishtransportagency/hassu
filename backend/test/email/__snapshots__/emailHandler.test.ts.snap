// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emailHandler sendEmailsByToiminto sendApprovalMailsAndAttachments should send emails and attachments succesfully 1`] = `
Object {
  "args": Array [
    Object {
      "Bucket": "hassu-localstack-yllapito",
      "Key": "yllapito/tiedostot/projekti/2/aloituskuulutus/ILMOITUS TOIMIVALTAISEN VIRANOMAISEN KUULUTUKSESTA Marikan testiprojekti.pdf",
    },
    Object {
      "Bucket": "hassu-localstack-yllapito",
      "Key": "yllapito/tiedostot/projekti/2/aloituskuulutus/KUULUTUS SUUNNITTELUN ALOITTAMISESTA Marikan testiprojekti.pdf",
    },
    Object {
      "Bucket": "hassu-localstack-yllapito",
      "Key": "yllapito/tiedostot/projekti/2/aloituskuulutus/MEDDELANDE OM KUNGORELSE FRAN BEHORIG MYNDIGHET Marikas testprojekt.pdf",
    },
  ],
  "stub": "getObject",
}
`;

exports[`emailHandler sendEmailsByToiminto sendApprovalMailsAndAttachments should send emails and attachments succesfully 2`] = `
Array [
  Object {
    "subject": "Valtion liikenneväylien suunnittelu: Aloituskuulutus hyväksytty ELY/2/2022",
    "text": "Valtion liikenneväylien suunnittelu -järjestelmän projektin
Testiprojekti 2
aloituskuulutus on hyväksytty.
Voit tarkastella aloituskuulutusta osoitteessa https://localhost:3000/yllapito/projekti/2/aloituskuulutus
Saat tämän viestin, koska sinut on merkitty aloituskuulutuksen laatijaksi. Tämä on automaattinen sähköposti, johon ei voi vastata.",
    "to": "matti.meikalainen@vayla.fi",
  },
  Object {
    "attachments": Array [
      Object {
        "contentType": "application/pdf",
        "filename": "KUULUTUS SUUNNITTELUN ALOITTAMISESTA Marikan testiprojekti.pdf",
      },
    ],
    "subject": "Valtion liikenneväylien suunnittelu: Aloituskuulutus hyväksytty ELY/2/2022",
    "text": "Valtion liikenneväylien suunnittelu -järjestelmän projektin
Testiprojekti 2
aloituskuulutus on hyväksytty. Liitteenä aloituskuulutus PDF-tiedostona, muistathan viedä sen asiakirjanhallintaan.

Voit tarkastella aloituskuulutusta osoitteessa https://localhost:3000/yllapito/projekti/2/aloituskuulutus

Saat tämän viestin, koska sinut on merkitty aloituskuulutuksen projektipäälliköksi. Tämä on automaattinen sähköposti, johon ei voi vastata.",
    "to": Array [
      "pekka.projari@vayla.fi",
    ],
  },
  Object {
    "attachments": Array [
      Object {
        "contentType": "application/pdf",
        "filename": "ILMOITUS TOIMIVALTAISEN VIRANOMAISEN KUULUTUKSESTA Marikan testiprojekti.pdf",
      },
      Object {
        "contentType": "application/pdf",
        "filename": "MEDDELANDE OM KUNGORELSE FRAN BEHORIG MYNDIGHET Marikas testprojekt.pdf",
      },
    ],
    "cc": Array [
      "pekka.projari@vayla.fi",
    ],
    "subject": "Ilmoitus Uudenmaan ELY-keskuksen kuulutuksesta",
    "text": "Tämä viesti on lähetetty Valtion liikenneväylien suunnittelu -järjestelmästä, älä vastaa tähän viestiin.

Testiprojekti 2

Uudenmaan elinkeino-, liikenne- ja ympäristökeskus (ELY-keskus) ilmoittaa, että se julkaisee kuulutuksen, joka koskee otsikossa mainitun tiesuunnitelman ja maastotöiden aloittamista. Uudenmaan ELY-keskus saattaa asian tiedoksi julkisesti kuuluttamalla vähintään 30 päivän ajan siten, kuin julkisesta kuulutuksesta säädetään hallintolaissa, sekä julkaisemalla kuulutuksen vähintään yhdessä alueella yleisesti ilmestyvässä sanomalehdessä. (LjMTL 108§, HL 62 a §)

Kuulutus julkaistaan 28.3.2022 osoitteessa https://localhost:3000/suunnitelma/2/aloituskuulutus.

Suunnittelualueen kuntia pyydetään julkaisemaan liitteenä oleva ilmoitus kuulutuksesta verkkosivuillaan. Ilmoitus tulee julkaista mahdollisuuksien mukaan edellä mainittuna kuulutuksen julkaisupäivänä. Ilmoituksen nähtävillä oloa ei tarvitse todentaa ilmoituksen lähettäjälle. Ilmoitus menee ELY-keskuksen ja Väyläviraston kuulutussivulle automaattisesti Valtion liikenneväylien suunnittelu palvelun kautta.

Suunnittelu kohde on osa Hyvinkää-Hanko sähköistyshanketta, jossa toteutetaan myös tasoristeysten toimenpiteitä. Hyväksytyssä ratasuunnitelmassa kyseisessä kohdassa on hyväksytty suunnitelmaratkaisuna Kisan seisakkeen ja Leksvallin tasoristeysten poistaminen parantamalla Helmströmin tasoristeyksen kohdalle uusi tasoristeys. Nyt käynnistetään Rata-suunnitelman päivitys kyseisessä kohdassa ja suunnitellaan kaikkien kolmen tasoristeyksen poistaminen uudella Leksvallin ylikulkusillalla.

Uudenmaan ELY-keskus käsittelee suunnitelman laatimiseen liittyen tarpeellisia henkilötietoja. Halutessasi tietää tarkemmin väyläsuunnittelun tietosuojakäytänteistä tutustu verkkosivujen tietosuojaosioon, https://www.ely-keskus.fi/tietosuoja.

Lisätietoja antaa 

Väylävirasto, Marika Ojanen, puh. 0299878787, marika.ojanen@vayla.fi

Detta meddelande har skickats från Planering av statens trafikleder, vänligen svara inte på detta.

Marikas testprojekt

Närings-, trafik- och miljöcentralen (NTM-centralen) i Nyland meddelar om publiceringen av en kungörelse som gäller inledande av den vägplan som nämns i rubriken samt terrängarbeten. NTM-centralen i Nyland delger ärendet offentligt genom kungörelse under minst 30 dagar på det sätt som föreskrivs om offentlig kungörelse i förvaltningslagen samt genom att publicera kungörelsen i flera dagstidningar med allmän spridning på området. (108 § i lagen om trafiksystem och landsvägar och 62 a § i förvaltningslagen)

Kungörelsen publiceras 28.3.2022 på adressen https://localhost:3000/sv/suunnitelma/2/nahtavillaolo.

Kommunerna i planeringsområdet ombeds publicera det bifogade meddelandet om kungörelse på sin webbplats. Meddelandet ska i mån av möjlighet publiceras den ovan nämnda dagen då kungörelsen publiceras. Man behöver inte bekräfta för avsändaren att meddelandet är framlagt. Meddelandet kommer automatiskt upp på NTM-centralens och Trafikledsverkets kungörelsesida via tjänsten Planering av statens trafikleder. 

Designplatsen är en del av elektrifieringsprojektet Hyvinge-Hangö, som också kommer att genomföra plankorsningsåtgärder. I den fastställda spårplanen är borttagandet av tävlingsstoppet och Leksvallskorsningen genom förbättring av en ny plankorsning vid Helmströms plankorsning vid denna punkt godkänd som planlösning. En uppdatering av Spårplanen kommer nu att lanseras då och man planerar att ta bort alla tre plankorsningarna på den nya Leksvallsöverfarten.

NTM-centralen i Nyland behandlar personuppgifter som är nödvändiga för utarbetandet av planen. Om du vill veta mer om trafikledsplaneringens dataskyddspolicy, bekanta dig med sektionen om dataskydd på webbplatsen https://www.ely-keskus.fi/tietosuoja

Mer information om planen 

Väylävirasto, Marika Ojanen, puh. 0299878787, marika.ojanen@vayla.fi",
    "to": Array [
      "mikkeli@mikke.li",
      "juva@ju.va",
      "savonlinna@savonlin.na",
      "kirjaamo.etela-savo@ely-keskus.fi",
    ],
  },
]
`;
